pkg/multishare_reconciler/reconciler.go

```diff
@@ -0,0 +1,873 @@
+		newStatus.Error = shareInfoClone.Status.Error
+	}
+	shareInfoClone.Status = newStatus
+	klog.Infof("trying to update ShareInfo %q status to %v", shareInfo.Name, shareInfoClone.Status)
+
+	return recon.updateShareInfoStatus(context.TODO(), shareInfoClone)
+}
+
+func (recon *multishareReconciler) createInstanceInfo(ctx context.Context, instanceInfo *v1alpha1.InstanceInfo) (*v1alpha1.InstanceInfo, error) {
+	klog.Infof("trying to create instanceInfo %s", instanceInfo.Name)
+	result, err := recon.clientset.MultishareV1alpha1().InstanceInfos().Create(context.TODO(), instanceInfo, metav1.CreateOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+func (recon *multishareReconciler) deleteInstanceInfo(ctx context.Context, instanceInfo *v1alpha1.InstanceInfo) error {
+	if len(instanceInfo.Finalizers) == 0 {
+		return fmt.Errorf("need to have finalizer to prevent auto gc of instanceInfo object")
+	}
+	klog.Infof("trying to add deletionTimestamp to instanceInfo %s", instanceInfo.Name)
+	return recon.clientset.MultishareV1alpha1().InstanceInfos().Delete(context.TODO(), instanceInfo.Name, metav1.DeleteOptions{})
+}
+
+func (recon *multishareReconciler) updateShareInfoStatus(ctx context.Context, shareInfoClone *v1alpha1.ShareInfo) (*v1alpha1.ShareInfo, error) {
+	result, err := recon.clientset.MultishareV1alpha1().ShareInfos().UpdateStatus(ctx, shareInfoClone, metav1.UpdateOptions{})
+	if err != nil {
+		return result, err
+	}
+	return result, nil
+}
+
+func (recon *multishareReconciler) updateInstanceInfoStatus(ctx context.Context, instanceInfoClone *v1alpha1.InstanceInfo) (*v1alpha1.InstanceInfo, error) {
+	result, err := recon.clientset.MultishareV1alpha1().InstanceInfos().UpdateStatus(ctx, instanceInfoClone, metav1.UpdateOptions{})
+	if err != nil {
+		return result, err
+	}
+	return result, nil
+}
+
+func (recon *multishareReconciler) updateInstanceInfo(ctx context.Context, instanceInfoClone *v1alpha1.InstanceInfo) (*v1alpha1.InstanceInfo, error) {
+	result, err := recon.clientset.MultishareV1alpha1().InstanceInfos().Update(ctx, instanceInfoClone, metav1.UpdateOptions{})
+	if err != nil {
+		return result, err
+	}
+	return result, nil
+}
+
+// TODO: unit testing this function
+// managedInstanceAndShare filters out instances and shares that are not managed by current cluster
+// the returned values should be treated as read only
+func (recon *multishareReconciler) managedInstanceAndShare(instances []*file.MultishareInstance, shares []*file.Share) ([]*file.MultishareInstance, []*file.Share, map[string][]*file.Share) {
+	var managedInstances []*file.MultishareInstance
+	var managedShares []*file.Share
+
+	instanceShare := make(map[string][]*file.Share)
+	clusterLocation := recon.cloud.Zone
+	if recon.config.IsRegional {
+		var err error
+		clusterLocation, err = util.GetRegionFromZone(clusterLocation)
+		if err != nil {
+			klog.Fatalf("failed to get region for regional cluster: %v", err.Error())
```

Comment:
let's return error from internal functions, and decide on the error handling in the caller

Reply:
done